### ISSUE TITLE
Null Checking

### ISSUE THREAD

**Issue by StevenThuriot** [DedAnton/NextGenMapper#114]
**Is your feature request related to a problem? Please describe.**

The generated mappers currently don't check for null references, which may cause exceptions if not every property is initialized. This might also be the case for nested mappings. Consider:

```csharp
record A(IEnumerable<B> list);
record B(string value);

record C(IEnumerable<D> list);
record D(string value);

List<B> bs = new(1) { new(null) };
A a = new(bs);

C c = a.MapTo<C>();
```

This will case an exception cause B's `string value` is null, but will also cause an exception already if we initialized A like this: `A a = new(null);`

Additionally, using nullable reference types seem to cause issues generating (couldn't get it to work). (e.g. `record E(string? value);`

**Describe the solution you'd like**

Adding a simple null check before calling a nested `.Map<T>`

e.g. the generated code would be something like this for `C c = a.MapTo<C>();`

Current:
```csharp
C MapTo<C>(this A source) => new C(source.list.MapTo<D>());
```

To be:
```csharp
C MapTo<C>(this A source) => new C(source.list?.MapTo<D>());
```
(Notice the additional question mark).

If the source is nullable, and the destination is not, `?? default(Destination)` should probably be added too.

This will be better for perfo than the other solution, as you'll only be doing null checks when absolutely necessary. Otherwise the user of the lib might also end up doing null checks twice (once themselves, once in the generated code) in the originally mentioned way.

This should be fairly easy to do by editing `ConfiguredMapsSourceBuilder` L157 to:
```csharp
                _builder.Append($"source.{propertyMap.SourceName}{(propertyMap.IsSourceNullable ? "?" : "")}.Map<{propertyMap.DestinationType}>()");

                if (propertyMap.IsSourceNullable && !propertyMap.IsDestinationNullable)
                {
                    _builder.Append($" ?? default({propertyMap.DestinationType})");
                }
```

and L176 to 
```csharp 
                _builder.Append($"{propertyMap.DestinationName} = source.{propertyMap.SourceName}{(propertyMap.IsSourceNullable ? "?" : "")}.Map<{propertyMap.DestinationType}>()");

                if (propertyMap.IsSourceNullable && !propertyMap.IsDestinationNullable)
                {
                    _builder.Append($" ?? default({propertyMap.DestinationType})");
                }
```

**Describe alternatives you've considered**

Since the mappers are nested, the easiest solution seems to be to include a null check in the mapping method if the value is nullable and return `default` if it is.

The easiest place to do this seems to be in `ConfiguredMapsSourceBuilder` :

```csharp
    private void ConstructorInvocation(ConfiguredMap map)
    {
        _builder.Append("\r\n        => ");

        if (map.IsSourceNullable)
        {
            _builder.Append($"ReferenceEquals(source, null) ? default({map.Destination}) : ");
        }

        _builder.Append($"new {map.Destination}");
```

**Additional context**

If there's another way to keep from crashing without too much overhead, I'd also be interested.

**Comment by DedAnton** [comment]
@StevenThuriot Thank for your request and your time spent investigation. 

Unfortunately, I cannot fully accept your proposal, because it contradicts the concept of my mapper. I was categorically against the fact that the mapper swallowed errors without telling the user anything (like AutoMapper). Therefore, I do not check for null anywhere in the generated methods for mapping.
I understand that this can cause an exception, but I think that swallowing an exception can become a much bigger problem and lead to a bug that is hard to find.

For nullable reference types, the situation is different because the user explicitly states that they expect to see a null value in the property. Therefore, I will add a check for null when calling the Map function for nullable properties (If both, source and destination properties are nullable annotated).
For other cases with nullable reference types, the mapper provides a diagnostic (NGM022 Possible null reference) that should tell the user about the problem.

I'm sorry you didn't manage to use the mapper with nullable reference types in your case, it was probably a bug. I would be grateful if you could provide more details and an example.

**Comment by DedAnton** [comment]
issue #116 (about null check for nullable annotated properties) was planned for next release

**Comment by StevenThuriot** [comment]
That's very fair and I can understand that. Definitely not a fan of swallowing the exceptions, though I'm not sure if I 100% agree that checking nulls in this case can be counted as such.

In my case, the contracts are third party and thus out of my control. Sadly they do not use nullable reference types yet, making it hard to use this for mapping without manually checking and initializing everything myself first to avoid null references. Doing all of these checks is almost as much work as just writing the mapping myself without the additional memory pressure of constantly initializing fields that can be null. While a very cool feature, I feel like the adoption rate is still quite low 😞 Especially in projects that existed before the feature came out.  A second, null-checking `Map` method might already be useful to have for these scenarios.

As for the example: 

This did not generate any mappings for me due to `NGM022	Possible null reference exception when mapping 'A.list' of type 'System.Collections.Generic.IEnumerable<B>?' to 'D.list' of type 'System.Collections.Generic.IEnumerable<E>?'`:

(though it seems like your next release will fix this)

```csharp
A a = new(new List<B>(1) { new(null) });

D d = a.Map<D>();

public record A(IEnumerable<B>? list);
public record B(C? value);
public record C(string value);

public record D(IEnumerable<E>? list);
public record E(F? value);
public record F(string value);
```

Changing it to this:

```csharp
A a = new(new List<B>(1) { new(null) });

D d = a.Map<D>();

public record A(IEnumerable<B> list);
public record B(C? value);
public record C(string value);

public record D(IEnumerable<E> list);
public record E(F? value);
public record F(string value);
```

Gave an `NGM022	Possible null reference exception when mapping 'B.value' of type 'C?' to 'E.value' of type 'F?'`. So I did have a map from `A` to `B` but not from `B` to `E` so it still did not compile. For obvious reasons, it did work if I removed all `?` but then I ended up with a nullref since `C` was `null`.

Additionally, it also gives an `NGM008	Mapping function for mapping 'B' to 'E' was not found`, but I'm assuming it's a direct result of the `NGM022` one.
